spotify: update sendPong, handle MP3_160_ENC

theSmallNothing opened the pull request:
- removes the dependency on http://git.io/WyCx0Q
- handles MP3_160_ENC format files

---

Had a bit of spare time over the holidays to look into this. Apologies for the wall of code (courtesy of Google Closure), it was either that or ~350 lines of pretty random assignments. Anyone manages to cut it down let me know, I'd love to know how.

dcarr622 committed:
new getStreamURL method for track

dcarr622 committed:
fix merge

dcarr622 committed:
Merge branch 'master' of github.com:TooTallNate/node-spotify-web

peetahzee committed:
Pass in a Callback function to get URL

peetahzee committed:
FacebookLogin conveniecne function

peetahzee committed:
up version

dcarr622 committed:
update ping pong url

dcarr622 committed:
log trackuri failure

dcarr622 committed:
use rootlist instead of publishedrootlist

dcarr622 committed:
update pingpong

dcarr622 committed:
handle pong response correctly

dcarr622 committed:
Merge branch 'master' of https://github.com/TooTallNate/node-spotify-web

dcarr622 committed:
Merge branch 'master' of https://github.com/TooTallNate/node-spotify-web

dcarr622 committed:
remove speaker

fuzeman committed:
Disabled SNI on websocket connections to fix #111

fuzeman committed:
Disabled SNI on websocket connections to fix #111

dcarr622 committed:
remove extra character

fuzeman committed:
Updated metadata protocol buffer and request version

#111

fuzeman committed:
Changed client `userAgent`

fuzeman committed:
Updated track restriction parsing

#111

fuzeman committed:
The `Track` get/metadata method didn't work correctly with protobuf.js

#111

dcarr622 committed:
merge fuzeman fixes

fuzeman committed:
Fixed bug with restriction country checks

#111

dcarr622 committed:
Merge remote-tracking branch 'fuzeman/issue/111' into fuzeman-fixes

dcarr622 committed:
check for res before res.uri

dcarr622 committed:
don't log null trackuri

dcarr622 committed:
fix track uri null handling

dcarr622 committed:
correct ping endpoint

dcarr622 committed:
fix pong response

dcarr622 committed:
use rootlist instead of publishedrootlist

longle255 committed:
Merge pull request #1 from sciencepro/master

merge from sciencepro

romualdr committed:
fixed ping/pong

Thanks to @fuzeman

bogsen committed:
Add spotify.library

mattyway committed:
Merge remote-tracking branch 'fuzeman/issue/111'

mattyway committed:
Updated metadata proto fields

mattyway committed:
Removed optional protobuf package

Commit:
Merge pull request #1 from Bogdacutu/patch-1

Patch 1

Commit:
update sendPong

removes the dependency on http://git.io/WyCx0Q

LinusU commented:
Awesome!

How did you manage to reverse engineer it?

TooTallNate commented:
Pretty badass 😎

Commit:
Merge branch 'master' of https://github.com/theSmallNothing/node-spotify-web into theSmallNothing-master

# Conflicts:
#	lib/spotify.js

Commit:
Merge pull request #4 from Shigawire/theSmallNothing-master

The small nothing master

Commit:
Intercept errors if spotify web service is down

ghost commented:
Cheers guys, @LinusU JPEXS now supporting Alchemy opcode decompilation helped, it would have been a lot more painful to work with the control flow on an unending stream of bytecodes. That cross-referenced with assembly of C++ stdlib allowed all the inlined functions to be removed, leaving what you see in the commit as the main algorithm. Weirdly, there did seem to be a second encryption routine in there that wasn't used, might be something to keep an eye on in the future.

Commit:
add new image key sizes for new metadata protobuf

ghost commented:
Haven't done extensive testing yet, but this appears to work fine! Very nice work

filfat commented:
Bump, LGTM?

Commit:
spotify: handle MP3_160_ENC

add lib/crypto.js: handles all key generation and decryption functionality
modify lib/spotify.js: handles sneaky `album_art` messages
modify lib/track.js: `play` now includes decryption transformer in pipeline
modify lib/util.js: added hex2bin and bin2hex helper functions
modify package.json: commit changes require newer `superagent`

Commit:
spotify: remove unmaintained protobuf dependency

modify package.json: remove references to `protobuf`
modify lib/schemas.js: remove references to `protobuf`, make `protobufjs` the only parser
remove proto/*.desc: only required for the `protobuf` module

Commit:
Merge branch 'master' of git://github.com/mattyway/node-spotify-web

Lordmau5 committed:
Fix for trying to fetch a whole album / playlist at once.

So as it turns out, the old code seems to have taken the very first "album_art" message over to every single track that followed, if an Album or Playlist was fetched.
This code basically runs it through a new Context within the code-execution VM every time one of those messages is being received.

Lordmau5 committed:
Fix library method not being able to fetch saved tracks

Lordmau5 committed:
Fix indentation

Commit:
Merge pull request #1 from Lordmau5/master

Fix for trying to fetch a whole album / playlist at once.
Fix library method not being able to fetch saved tracks.

RealZyXeL commented:
Is it possible to use part of this code to decrypt 160kb/s MP3s from the Spotify Web-Player or did they changed something in the meantime?

denysvitali commented:
I've already tried without success unfortunately. If somebody finds a way to do it cc me please :)

RealZyXeL commented:
Strange... The swf key is definitly the same as before - this was the reason I found this website ;-)

denysvitali commented:
I know, but apparently the resulting file is garbage. They may have changed something, or it is me that I'm not good at implementing stuff
